[WIP] New shard proposal (#1427)
ethereum/eth2.0-specs: specs/core/1_new_shards.md

    Takes as input balance-in-increments (// EFFECTIVE_BALANCE_INCREMENT) to preserve symmetry with
    the unpacking function.
    """
    return (index << 16) + (slashed << 15) + balance_in_increments
```

### `committee_to_compact_committee`

```python
def committee_to_compact_committee(state: BeaconState, committee: Sequence[ValidatorIndex]) -> CompactCommittee:
    """
    Given a state and a list of validator indices, outputs the CompactCommittee representing them.
    """
    validators = [state.validators[i] for i in committee]
    compact_validators = [
        pack_compact_validator(i, v.slashed, v.effective_balance // EFFECTIVE_BALANCE_INCREMENT)
        for i, v in zip(committee, validators)
    ]
    pubkeys = [v.pubkey for v in validators]
    return CompactCommittee(pubkeys=pubkeys, compact_validators=compact_validators)
```

### `get_light_client_committee`

```python
def get_light_client_committee(beacon_state: BeaconState, epoch: Epoch) -> Sequence[ValidatorIndex]:
    assert epoch % LIGHT_CLIENT_COMMITTEE_PERIOD == 0
    active_validator_indices = get_active_validator_indices(beacon_state, epoch)

>     seed = get_seed(beacon_state, epoch, DOMAIN_SHARD_LIGHT_CLIENT)

Author: That's true though we have expressed desire for more domain separation before. I'd be ok either way.

>     seed = get_seed(beacon_state, epoch, DOMAIN_SHARD_LIGHT_CLIENT)

Reviewer: Separate does feel more in line with past decisions. Keep as is

>     seed = get_seed(beacon_state, epoch, DOMAIN_SHARD_LIGHT_CLIENT)

Reviewer: Separate does feel more in line with past decisions. Keep as is

```

### `get_indexed_attestation`

```python
def get_indexed_attestation(beacon_state: BeaconState, attestation: Attestation) -> IndexedAttestation:
    attesting_indices = get_attesting_indices(state, attestation.data, attestation.aggregation_bits)
    return IndexedAttestation(attesting_indices, data, custody_bits, signature)
```

### `update_gasprice`

```python
def update_gasprice(prev_gasprice: Gwei, length: uint8) -> Gwei:
    if length > BLOCK_SIZE_TARGET:
        delta = prev_gasprice * (length - BLOCK_SIZE_TARGET) // BLOCK_SIZE_TARGET // GASPRICE_ADJUSTMENT_COEFFICIENT
        return min(prev_gasprice + delta, MAX_GASPRICE)
    else:
        delta = prev_gasprice * (BLOCK_SIZE_TARGET - length) // BLOCK_SIZE_TARGET // GASPRICE_ADJUSTMENT_COEFFICIENT
        if delta > prev_gasprice - GASPRICE_ADJUSTMENT_COEFFICIENT:
            return GASPRICE_ADJUSTMENT_COEFFICIENT
        else:
            return prev_gasprice - delta
```

### `is_valid_indexed_attestation`

```python
def is_valid_indexed_attestation(state: BeaconState, indexed_attestation: IndexedAttestation) -> bool:
    """
    Check if ``indexed_attestation`` has valid indices and signature.
    """

> 

Author: The length verification is already done by the SSZ datatype.

> 

Reviewer: right

    # Verify indices are sorted
    if indexed_attestation.participants != sorted(indexed_attestation.participants):
        return False
    
    # Verify aggregate signature
    all_pubkeys = []
    all_message_hashes = []

>     for participant, custody_bits in zip(participants, indexed_attestation.custody_bits):

Author: The problem is that in the normal case, the number of blocks in the attestation is ~1-4, so making a list of length 1-4 bitlists would be a huge amount of hashing overhead. But I fixed the loops to do things in the right order, good catch.

>     for participant, custody_bits in zip(participants, indexed_attestation.custody_bits):

Reviewer: gotcha. looks good

        for i, bit in enumerate(custody_bits):
            all_pubkeys.append(state.validators[participant].pubkey)
            # Note: only 2N distinct message hashes
            all_message_hashes.append(AttestationCustodyBitWrapper(hash_tree_root(indexed_attestation.data), i, bit))
        
    return bls_verify_multiple(
        pubkeys=all_pubkeys,
        message_hashes=all_message_hashes,
        signature=indexed_attestation.signature,
        domain=get_domain(state, DOMAIN_BEACON_ATTESTER, indexed_attestation.data.target.epoch),
    )
```

## Beacon Chain Changes

### New state variables

```python
    shard_states: Vector[ShardState, MAX_SHARDS]

>     online_countdown: Bytes[VALIDATOR_REGISTRY_LIMIT]

Author: I was keeping it as a list of bytes because that way the hashing is lower. Hashing over a uint64 for every validator is currently only done in the validator balances list and is a primary component of epoch transition overhead; adding yet more hashing of this type would significantly add overhead.

If we think using bytes/uint8 is too impure, I have another suggestion: have a bitlist "online during previous 4-epoch period" and "online during current 4-epoch period". Validators would be considered online if their bit if flipped on for either one. This would have even less hashing overhead and would use existing types.

>     online_countdown: Bytes[VALIDATOR_REGISTRY_LIMIT]

Reviewer 2: I think List[uint8, ...] is clearer for this purpose than Bytes[...]

    current_light_committee: CompactCommittee
    next_light_committee: CompactCommittee
```

### New block data structures

```python
    shard_transitions: Vector[ShardTransition, MAX_SHARDS]
    light_client_signature_bitfield: Bitlist[LIGHT_CLIENT_COMMITTEE_SIZE]
    light_client_signature: BLSSignature
```

### Attestation processing

```python
def process_attestation(state: BeaconState, attestation: Attestation) -> None:
    data = attestation.data
    assert data.index < ACTIVE_SHARDS
    shard = (data.index + get_start_shard(state, data.slot)) % ACTIVE_SHARDS
    proposer_index=get_beacon_proposer_index(state)

    # Signature check
    committee = get_crosslink_committee(state, get_current_epoch(state), shard)
    for bits in attestation.custody_bits + [attestation.aggregation_bits]:
        assert len(bits) == len(committee)
    # Check signature
    assert is_valid_indexed_attestation(state, get_indexed_attestation(state, attestation))
    # Get attesting indices
    attesting_indices = get_attesting_indices(state, attestation.data, attestation.aggregation_bits)

    # Prepare pending attestation object

>     pending_attestation = PendingAttestation(

Reviewer: pending_attestation already has `data` as a field so no need to add `slot` or `shard` explicitly

        slot=data.slot,
        shard=shard,
        aggregation_bits=attestation.aggregation_bits,
        inclusion_delay=state.slot - data.slot,
        crosslink_success=False,
        proposer_index=proposer_index
    )
    
    # Type 1: on-time attestations
    if data.custody_bits != []:
        # Correct slot
        assert data.slot == state.slot
        # Slot the attestation starts counting from
        start_slot = state.shard_next_slots[shard]
        # Correct data root count
        offset_slots = [start_slot + x for x in SHARD_BLOCK_OFFSETS if start_slot + x < state.slot]
        assert len(attestation.custody_bits) == len(offset_slots)
        # Correct parent block root
        assert data.beacon_block_root == get_block_root_at_slot(state, state.slot - 1)
        # Apply
        online_indices = get_online_indices(state)

>         if get_total_balance(state, online_indices.intersection(attesting_indices)) * 3 >= get_total_balance(state, online_indices) * 2:

Reviewer: Interesting implication of this:

In the previous design, attestations that were for the same crosslink data but were not aggregatable due to some amount of overlapping indices could both be included and count towards crosslinks and finality.

In this design, they _must_ be aggregated to both count toward the crosslinking, whereas non-aggregated can still both go toward finality.

Reviewer: This seems fine, but puts a larger reliance on good, non-overlapping aggregation

Author: Interesting! We should discuss this; definitely open to considering going back to the "loop through attestations first, then finish in a second round" option.

>             # Check correct formatting of shard transition data

Reviewer: Need to consider what happens if a fully duplicate attestation (or a not fully duplicate attestation, but both have 2/3 voting on a crosslink) is included. This would create two `PendingAttestations` with `crosslink_success` as `True`

Reviewer: Depending on how rewards are done, it is likely fine. Just need to get the union of the successful participants at `slot/index`.

            for i in range(len(offset_slots)):
                shard_state, block_length, chunks = transition.shard_states[i], transition.shard_block_lengths[i], transition.shard_data_roots[i]
                block_length = transition.shard
                assert shard_state.gasprice == update_gasprice(prev_gasprice, block_length)
                assert shard_state.slot == offset_slots[i]
                assert len(chunks) == block_length // SHARD_BLOCK_CHUNK_SIZE
                filled_roots = chunks + [EMPTY_CHUNK_ROOT] * (MAX_SHARD_BLOCK_CHUNKS - len(chunks))
                assert shard_state.latest_block_hash == hash_tree_root(filled_roots)
                prev_gasprice = shard_state.gasprice

            # Save updated state
            state.shard_states[shard] = transition.shard_states[-1]
            state.shard_states[shard].slot = state.slot - 1

            # Save success (for end-of-epoch rewarding)
            pending_attestation.crosslink_success = True

            # Apply proposer reward and cost
            estimated_attester_reward = sum([get_base_reward(state, attester) for attester in attesting_indices])

>             increase_balance(state, proposer, estimated_attester_reward // PROPOSER_REWARD_COEFFICIENT)

Reviewer: Why not do this reward during epoch transition?

Reviewer: Is this an _additional_ reward for successful crosslink inclusion?
Rather than the attestation inclusion reward that already exists in phase 0?

Author: I suppose we could, though if we do that we would also need to thread through the per-byte block fee.

            for state, length in zip(transition.shard_states, transition.shard_block_lengths):

>                 decrease_balance(state, proposer, state.gasprice * length)

Reviewer: Why is there a cost for including attestations? We don't currently have a `state.gasprice`.
Should this instead be a cost to the shard proposers?

            
    # Type 2: delayed attestations
    else:

>         assert slot_to_epoch(data.slot) in (get_current_epoch(state), get_previous_epoch(state))

Reviewer: Should this be within the last `SLOTS_PER_EPOCH` slots? That is how it is in phase 0 so there aren't asymmetries depending on which slot you are assigned to

        assert data.shard_transition_hash == Hash()
        assert len(attestation.custody_bits) == 0

    for index in attesting_indices:
        online_countdown[index] = ONLINE_PERIOD

    if data.target.epoch == get_current_epoch(state):
        assert data.source == state.current_justified_checkpoint
        state.current_epoch_attestations.append(pending_attestation)
    else:
        assert data.source == state.previous_justified_checkpoint
        state.previous_epoch_attestations.append(pending_attestation)
```

### Misc block post-processing

```python
def misc_block_post_process(state: BeaconState, block: BeaconBlock):
    # Verify that a `shard_transition` in a block is empty if an attestation was not processed for it
    for shard in range(MAX_SHARDS):
        if state.shard_states[shard].slot != state.slot - 1:
            assert block.shard_transition[shard] == ShardTransition()
```

### Light client processing

```python
def verify_light_client_signatures(state: BeaconState, block: BeaconBlock):
    period_start = get_current_epoch(state) - get_current_epoch(state) % LIGHT_CLIENT_COMMITTEE_PERIOD

>     committee = get_light_client_committee(state, period_start - min(period_start, LIGHT_CLIENT_COMMITTEE_PERIOD))

Reviewer: Why the `min` here?

Shouldn't it just be:

```suggestion
    committee = get_light_client_committee(state, period_start)
```

Author: We need to have logic somewhere to say "the light client committee used at slot N takes data from slot N - N % PERIOD - PERIOD (with one special case where N < PERIOD to avoid underflow). I just fixed; moved all that logic into `get_light_client_committee` itself.

>     signer_keys = []

Reviewer: Should add an `assert` that the `light_client_signature_bitfield` is of the expected length-- `len(committee)`

    for i, bit in enumerate(block.light_client_signature_bitfield):
        if bit:
            signer_keys.append(state.validators[committee[i]].pubkey)

>             increase_balance(state, committee[i], get_base_reward(state, committee[i]))

Reviewer: Seems like we need a proposer reward here

```

### Fraud proofs

TODO. The intent is to have a single universal fraud proof type, which contains (i) an on-time attestation on shard `s` signing a set of `data_roots`, (ii) an index `i` of a particular data root to focus on, (iii) the full contents of the i'th data, (iii) a Merkle proof to the `shard_state_roots` in the parent block the attestation is referencing, and which then verifies that one of the two conditions is false:

* `custody_bits[i][j] != generate_custody_bit(subkey, block_contents)` for any `j`
* `execute_state_transition(shard, slot, attestation.shard_state_roots[i-1], hash_tree_root(parent), get_shard_proposer(state, shard, slot), block_contents) != shard_state_roots[i]` (if `i=0` then instead use `parent.shard_state_roots[s][-1]`)

## Shard state transition function

```python
def shard_state_transition(shard: Shard, slot: Slot, pre_state: Hash, previous_beacon_root: Hash, proposer_pubkey: BLSPubkey, block_data: BytesN[MAX_SHARD_BLOCK_CHUNKS * SHARD_BLOCK_CHUNK_SIZE]) -> Hash:
    # Beginning of block data is the previous block hash

>     assert block_data[:32] == pre_state.latest_block_hash

Reviewer: Why not a more structured shard block with header fields and a data field?

    assert block_data[32:64] == int_to_bytes8(slot) + b'\x00' * 24
    # Signature check
    assert len(block_data) >= 160
    assert bls_verify(
        pubkey=proposer_pubkey,
        message_hash=hash_tree_root(block_data[:-96]),

>         signature=block_data[-96:],

Reviewer: again, why the unstructured-ness?

        domain=DOMAIN_SHARD_PROPOSER
    )
    # We will add something more substantive in phase 2

>     length = len(block.data.rstrip(b'\x00'))

Reviewer: `block` is not defined

    return ShardState(slot=slot, root=hash(pre_state + hash_tree_root(block_data)), gasprice=update_gasprice(pre_state, length), latest_block_hash = hash(block_data))
```

## Honest committee member behavior

> Suppose you are a committee member on shard `shard` at slot `current_slot`. Let `state` be the head beacon state you are building on. Three seconds into slot `slot`, run the following procedure:

Reviewer: Dividing the slot into 4

- beacon block proposal
- shard block proposal
- attestation broadcast
- aggregate broadcast

```suggestion
Suppose you are a committee member on shard `shard` at slot `current_slot`. Let `state` be the head beacon state you are building on. `2 * SECONDS_PER_SLOT // 4` seconds into slot `slot`, run the following procedure:
```

    - [Honest committee member behavior](#honest-committee-member-behavior)

<!-- /TOC -->

## Introduction

This document describes the shard transition function (data layer only) and the shard fork choice rule as part of Phase 1 of Ethereum 2.0.

## Configuration

### Misc

| Name | Value | Unit | Duration |
| - | - | - | - | 
| `MAX_SHARDS` | `2**10` (= 1024) |
| `ACTIVE_SHARDS` | `2**6` (= 64) |
| `ONLINE_PERIOD` | `2**3` (= 8) | epochs | ~51 min |
| `LIGHT_CLIENT_COMMITTEE_SIZE` | `2**7` (= 128) |
| `LIGHT_CLIENT_COMMITTEE_PERIOD` | `2**8` (= 256) | epochs | ~27 hours |
| `SHARD_BLOCK_CHUNK_SIZE` | `2**18` (= 262,144) | |
| `MAX_SHARD_BLOCK_CHUNKS` | `2**2` (= 4) | |
| `BLOCK_SIZE_TARGET` | `3 * 2**16` (= 196,608) | |
| `SHARD_BLOCK_OFFSETS` | `[1, 2, 3, 5, 8, 13, 21, 34, 55, 89, 144, 233]` | |
| `MAX_SHARD_BLOCKS_PER_ATTESTATION` | `len(SHARD_BLOCK_OFFSETS)` | |
| `EMPTY_CHUNK_ROOT` | `hash_tree_root(BytesN[SHARD_BLOCK_CHUNK_SIZE]())` | |
| `MAX_GASPRICE` | `2**14` (= 16,384) | Gwei | |
| `GASPRICE_ADJUSTMENT_COEFFICIENT` | `2**3` (= 8) | |
| `DOMAIN_SHARD_LIGHT_CLIENT` | `192` | |
| `DOMAIN_SHARD_PROPOSAL` | `193` | |

## Containers

### `ShardBlockWrapper`

```python

> class ShardBlockWrapper(Container):

Reviewer 3: I suppose there will also have `ShardBlock`?

    shard_parent_root: Hash
    beacon_parent_root: Hash
    slot: Slot
    body: BytesN[SHARD_BLOCK_CHUNK_SIZE]
    signature: BLSSignature
```

### `ShardSignedHeader`

```python

> class ShardSignedHeader(Container):

Reviewer 3: Could you elaborate on what does "signed" mean here?

    shard_parent_root: Hash
    beacon_parent_root: Hash
    slot: Slot
    body_root: Hash
```

### `ShardState`

```python
class ShardState(Container):
    slot: Slot

>     gasprice: Gwei

Reviewer 3: Need to set initial `gasprice` value.

Author: I think no need. It's zero by default, but then after one slot `get_updated_gasprice` pushes it up to `MIN_GASPRICE`.

    root: Hash
    latest_block_hash: Hash
```

### `AttestationData`

```python
class AttestationData(Container):
    slot: Slot
    index: CommitteeIndex
    # LMD GHOST vote
    beacon_block_root: Hash
    # FFG vote
    source: Checkpoint
    target: Checkpoint
    # Shard transition root
    shard_transition_root: Hash
```

### `ShardTransition`

```python
class ShardTransition(Container):
    # Starting from slot
    start_slot: Slot
    # Shard block lengths
    shard_block_lengths: List[uint64, MAX_SHARD_BLOCKS_PER_ATTESTATION]
    # Shard data roots
    shard_data_roots: List[List[Hash, MAX_SHARD_BLOCK_CHUNKS], MAX_SHARD_BLOCKS_PER_ATTESTATION]

>     # Intermediate state roots

Reviewer 3: ```suggestion
    # Intermediate shard states
```

>     shard_state_roots: List[ShardState, MAX_SHARD_BLOCKS_PER_ATTESTATION]

Reviewer 3: Should it be renamed to `shard_states`?
```suggestion
    shard_states: List[ShardState, MAX_SHARD_BLOCKS_PER_ATTESTATION]
```

    # Proposer signature aggregate
    proposer_signature_aggregate: BLSSignature
```

### `Attestation`

```python
class Attestation(Container):
    aggregation_bits: Bitlist[MAX_VALIDATORS_PER_COMMITTEE]
    data: AttestationData
    custody_bits: List[Bitlist[MAX_VALIDATORS_PER_COMMITTEE], MAX_SHARD_BLOCKS_PER_ATTESTATION]
    signature: BLSSignature
```

### `IndexedAttestation`

```python
class IndexedAttestation(Container):
    committee: List[ValidatorIndex, MAX_VALIDATORS_PER_COMMITTEE]

>     attestation: Attestation

Reviewer 3: It seems very different from [v09x `IndexedAttestation`](https://github.com/ethereum/eth2.0-specs/blob/v09x/specs/core/0_beacon-chain.md#indexedattestation), maybe give it a new name?

```

### `CompactCommittee`

```python
class CompactCommittee(Container):
    pubkeys: List[BLSPubkey, MAX_VALIDATORS_PER_COMMITTEE]
    compact_validators: List[uint64, MAX_VALIDATORS_PER_COMMITTEE]
```

### `AttestationCustodyBitWrapper`

```python
class AttestationCustodyBitWrapper(Container):
    attestation_root: Hash

>     block_index: uint64

Author: It's needed to distinguish the 0/1 values for each position, otherwise you'd just be adding up 1 bits without regard for what position they come in.

    light_client_signature_bitfield: Bitlist[LIGHT_CLIENT_COMMITTEE_SIZE]
    light_client_signature: BLSSignature
```

### Attestation processing

#### `validate_attestation`

```python
def validate_attestation(state: BeaconState, attestation: Attestation) -> None:
    data = attestation.data
    assert data.index < ACTIVE_SHARDS
    shard = get_shard(state, attestation)
    proposer_index = get_beacon_proposer_index(state)

    # Signature check
    assert is_valid_indexed_attestation(state, get_indexed_attestation(state, attestation))
    # Type 1: on-time attestations
    if data.custody_bits != []:
        # Correct slot
        assert data.slot == state.slot
        # Correct data root count
        start_slot = state.shard_next_slots[shard]
        offset_slots = [start_slot + x for x in SHARD_BLOCK_OFFSETS if start_slot + x < state.slot]
        assert len(attestation.custody_bits) == len(offset_slots)
        # Correct parent block root
        assert data.beacon_block_root == get_block_root_at_slot(state, state.slot - 1)
    # Type 2: delayed attestations
    else:
        assert state.slot - slot_to_epoch(data.slot) < EPOCH_LENGTH

>         assert data.shard_transition_root == Hash()

Reviewer 3: - Is it a kind of reduced form of attestation data?
- If so, does that mean the late attesters have to rebuild `AttestationData` of the old slot with empty `custody_bits` and `shard_transition_root`, and broadcast it?
- It seems the proposer can help transform ReducedAttestation in the [previous version](https://github.com/ethereum/eth2.0-specs/commit/0469d57ea26c848ffc2864741cc12d7dba2986c4)?

Author: Yes, it's a reduced form, and late attesters do have to rebroadcast.

        # The committee for this shard
        this_shard_committee = get_beacon_committee(state, get_current_epoch(state), shard)
        # Loop over all shard transition roots
        for shard_transition_root in sorted(set([attestation.data.shard_transition_root for attestation in this_shard_attestations])):
            all_participants = set()
            participating_attestations = []
            for attestation in this_shard_attestations:
                participating_attestations.append(attestation)
                if attestation.data.shard_transition_root == shard_transition_root:
                    all_participants = all_participants.union(get_attesting_indices(state, attestation.data, attestation.aggregation_bits))
            if (
                get_total_balance(state, online_indices.intersection(all_participants)) * 3 >=
                get_total_balance(state, online_indices.intersection(this_shard_committee)) * 2
                and success is False
            ):
                assert shard_transition_root == hash_tree_root(block.shard_transition)

>                 process_crosslink(state, shard, block.shard_transition)

Author: ```suggestion
                apply_shard_transition(state, shard, block.shard_transition)
```

| `SHARD_BLOCK_CHUNK_SIZE` | `2**18` (= 262,144) | |
| `MAX_SHARD_BLOCK_CHUNKS` | `2**2` (= 4) | |
| `BLOCK_SIZE_TARGET` | `3 * 2**16` (= 196,608) | |
| `SHARD_BLOCK_OFFSETS` | `[1, 2, 3, 5, 8, 13, 21, 34, 55, 89, 144, 233]` | |
| `MAX_SHARD_BLOCKS_PER_ATTESTATION` | `len(SHARD_BLOCK_OFFSETS)` | |
| `EMPTY_CHUNK_ROOT` | `hash_tree_root(BytesN[SHARD_BLOCK_CHUNK_SIZE]())` | |
| `MAX_GASPRICE` | `2**14` (= 16,384) | Gwei | |
| `MIN_GASPRICE` | `2**5` (= 32) | Gwei | |
| `GASPRICE_ADJUSTMENT_COEFFICIENT` | `2**3` (= 8) | |
| `DOMAIN_SHARD_LIGHT_CLIENT` | `192` | |
| `DOMAIN_SHARD_PROPOSAL` | `193` | |

## Containers

### `ShardBlockWrapper`

_Wrapper for being broadcasted over the network._

```python
class ShardBlockWrapper(Container):
    shard_parent_root: Hash
    beacon_parent_root: Hash
    slot: Slot
    body: BytesN[SHARD_BLOCK_CHUNK_SIZE]
    signature: BLSSignature
```

### `ShardSignableHeader`

```python
class ShardSignableHeader(Container):
    shard_parent_root: Hash
    beacon_parent_root: Hash
    slot: Slot
    body_root: Hash
```

### `ShardState`

```python
class ShardState(Container):
    slot: Slot
    gasprice: Gwei
    data: Hash

>     latest_block_hash: Hash

Reviewer 4: `block_root` instead of `block_hash`?

        - [Epoch transition](#epoch-transition)
    - [Fraud proofs](#fraud-proofs)
    - [Shard state transition function](#shard-state-transition-function)
    - [Honest committee member behavior](#honest-committee-member-behavior)

<!-- /TOC -->

## Introduction

This document describes the shard transition function (data layer only) and the shard fork choice rule as part of Phase 1 of Ethereum 2.0.

## Configuration

### Misc

| Name | Value | Unit | Duration |
| - | - | - | - | 
| `MAX_SHARDS` | `2**10` (= 1024) |
| `ACTIVE_SHARDS` | `2**6` (= 64) |
| `ONLINE_PERIOD` | `2**3` (= 8) | epochs | ~51 min |
| `LIGHT_CLIENT_COMMITTEE_SIZE` | `2**7` (= 128) |
| `LIGHT_CLIENT_COMMITTEE_PERIOD` | `2**8` (= 256) | epochs | ~27 hours |
| `SHARD_COMMITTEE_PERIOD` | `2**8` (= 256) | epochs | ~27 hours |
| `SHARD_BLOCK_CHUNK_SIZE` | `2**18` (= 262,144) | |
| `MAX_SHARD_BLOCK_CHUNKS` | `2**2` (= 4) | |
| `BLOCK_SIZE_TARGET` | `3 * 2**16` (= 196,608) | |
| `SHARD_BLOCK_OFFSETS` | `[1, 2, 3, 5, 8, 13, 21, 34, 55, 89, 144, 233]` | |
| `MAX_SHARD_BLOCKS_PER_ATTESTATION` | `len(SHARD_BLOCK_OFFSETS)` | |
| `EMPTY_CHUNK_ROOT` | `hash_tree_root(BytesN[SHARD_BLOCK_CHUNK_SIZE]())` | |

> | `MAX_GASPRICE` | `2**14` (= 16,384) | Gwei | |

Reviewer 4: `gasprice` or `gas_price`? not sure which is more semantic 🤔

| `MIN_GASPRICE` | `2**5` (= 32) | Gwei | |
| `GASPRICE_ADJUSTMENT_COEFFICIENT` | `2**3` (= 8) | |
| `DOMAIN_SHARD_LIGHT_CLIENT` | `192` | |
| `DOMAIN_SHARD_COMMITTEE` | `192` | |
| `DOMAIN_SHARD_PROPOSAL` | `193` | |

## Containers

### `ShardBlockWrapper`

_Wrapper for being broadcasted over the network._

```python
class ShardBlockWrapper(Container):
    shard_parent_root: Hash
    beacon_parent_root: Hash
    slot: Slot
    body: BytesN[SHARD_BLOCK_CHUNK_SIZE]
    signature: BLSSignature
```

### `ShardSignableHeader`

```python
class ShardSignableHeader(Container):
    shard_parent_root: Hash
    beacon_parent_root: Hash
    slot: Slot
    body_root: Hash
```

### `ShardState`

```python
class ShardState(Container):
    slot: Slot
    gasprice: Gwei
    data: Hash
    latest_block_root: Hash
```

### `AttestationData`

```python

> class AttestationData(Container):

Reviewer 4: Would it make sense to add `shard` to the field?  can avoid `get_shard`

Reviewer: Would still need to validate the shard against the committee index for the slot, but I see value in including it here explicitly (similar to including `slot` even though it used to be able to be calculated from the constituent parts)

    data: AttestationData
    custody_bits: List[Bitlist[MAX_VALIDATORS_PER_COMMITTEE], MAX_SHARD_BLOCKS_PER_ATTESTATION]
    signature: BLSSignature
```

### `AttestationAndCommittee`

```python
class AttestationAndCommittee(Container):
    committee: List[ValidatorIndex, MAX_VALIDATORS_PER_COMMITTEE]
    attestation: Attestation
```

### `CompactCommittee`

```python
class CompactCommittee(Container):
    pubkeys: List[BLSPubkey, MAX_VALIDATORS_PER_COMMITTEE]
    compact_validators: List[uint64, MAX_VALIDATORS_PER_COMMITTEE]
```

### `AttestationCustodyBitWrapper`

```python
class AttestationCustodyBitWrapper(Container):
    attestation_root: Hash
    block_index: uint64
    bit: bool
```

### `PendingAttestation`

```python
class PendingAttestation(Container):
    aggregation_bits: Bitlist[MAX_VALIDATORS_PER_COMMITTEE]
    data: AttestationData
    inclusion_delay: Slot
    proposer_index: ValidatorIndex
    crosslink_success: bool
```

## Helpers

### `get_online_validators`

```python
def get_online_indices(state: BeaconState) -> Set[ValidatorIndex]:
    active_validators = get_active_validator_indices(state, get_current_epoch(state))

>     return set([i for i in active_validators if state.online_countdown[i] != 0])

Reviewer 4: Is `set` really necessary?  Trying to optimize from client implementation's perspective

Reviewer 5: It's not, but when things are unordered and unique we declare it as `Set` in the spec. Here it would be a free conversion for a client to a set object anyway (or any validator indices container).

```

### `pack_compact_validator`

```python
def pack_compact_validator(index: int, slashed: bool, balance_in_increments: int) -> int:
    """
    Creates a compact validator object representing index, slashed status, and compressed balance.
    Takes as input balance-in-increments (// EFFECTIVE_BALANCE_INCREMENT) to preserve symmetry with
    the unpacking function.
    """
    return (index << 16) + (slashed << 15) + balance_in_increments
```

### `committee_to_compact_committee`

```python
def committee_to_compact_committee(state: BeaconState, committee: Sequence[ValidatorIndex]) -> CompactCommittee:
    """
    Given a state and a list of validator indices, outputs the CompactCommittee representing them.
    """

>     validators = [state.validators[i] for i in committee]

Reviewer 4: Can simplify to one loop without sacrificing much readability. Example:
```go
	compactValidators := make([]uint64, len(committee))
	pubKeys := make([][]byte, len(committee))

	for i := 0; i < len(committee); i++ {
		v := state.Validators[committee[i]]
		compactValidators[i] = PackCompactValidator(committee[i], v.Slashed, v.EffectiveBalance/params.BeaconConfig().EffectiveBalanceIncrement)
		pubKeys[i] = v.PublicKey
	}
```

    compact_validators = [
        pack_compact_validator(i, v.slashed, v.effective_balance // EFFECTIVE_BALANCE_INCREMENT)
        for i, v in zip(committee, validators)
    ]
    pubkeys = [v.pubkey for v in validators]
    return CompactCommittee(pubkeys=pubkeys, compact_validators=compact_validators)
```

### `get_shard_committee`

```python

> def get_shard_committee(beacon_state: BeaconState, epoch: Epoch, shard: Shard) -> Sequence[ValidatorIndex]:

Reviewer 4: `shard: Shard` is not used

    source_epoch = epoch - epoch % SHARD_COMMITTEE_PERIOD 

>     if source_epoch > 0:

Reviewer 4: This feels safer 

```suggestion
    if source_epoch >= SHARD_COMMITTEE_PERIOD:
```

        source_epoch -= SHARD_COMMITTEE_PERIOD
    active_validator_indices = get_active_validator_indices(beacon_state, source_epoch)
    seed = get_seed(beacon_state, source_epoch, DOMAIN_SHARD_COMMITTEE)
    return compute_committee(active_validator_indices, seed, 0, ACTIVE_SHARDS)
```

### `get_shard_proposer_index`

```python
def get_shard_proposer_index(beacon_state: BeaconState, slot: Slot, shard: Shard) -> ValidatorIndex:

>     committee = get_shard_committee(beacon_state, slot_to_epoch(slot), shard)

Reviewer 4: ```suggestion
    committee = get_shard_committee(beacon_state, compute_epoch_at_slot(slot), shard)
```

    all_message_hashes = []
    aggregation_bits = indexed_attestation.attestation.aggregation_bits
    assert len(aggregation_bits) == len(indexed_attestation.committee)
    for i, custody_bits in enumerate(indexed_attestation.attestation.custody_bits):
        assert len(custody_bits) == len(indexed_attestation.committee)
        for participant, abit, cbit in zip(indexed_attestation.committee, aggregation_bits, custody_bits):
            if abit:
                all_pubkeys.append(state.validators[participant].pubkey)
                # Note: only 2N distinct message hashes
                all_message_hashes.append(hash_tree_root(
                    AttestationCustodyBitWrapper(hash_tree_root(indexed_attestation.data), i, cbit)
                ))
            else:
                assert cbit == False
        
    return bls_verify_multiple(
        pubkeys=all_pubkeys,
        message_hashes=all_message_hashes,
        signature=indexed_attestation.signature,
        domain=get_domain(state, DOMAIN_BEACON_ATTESTER, indexed_attestation.data.target.epoch),
    )
```

### `get_shard`

```python
def get_shard(state: BeaconState, attestation: Attestation) -> Shard:

>     return Shard((attestation.data.index + get_start_shard(state, data.slot)) % ACTIVE_SHARDS)

Reviewer 4: `get_start_shard` is no longer defined in beacon chain spec, I think we have to reintroduce it here

>     return Shard((attestation.data.index + get_start_shard(state, data.slot)) % ACTIVE_SHARDS)

Reviewer 4: ```suggestion
    return Shard((attestation.data.index + get_start_shard(state, attestation.data.slot)) % ACTIVE_SHARDS)
```
